sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisReaderTest.java

```diff
@@ -119,4 +127,85 @@ public void readsThroughAllDataAvailable() throws IOException, TransientKinesisE
     assertThat(reader.advance()).isFalse();
   }
 
+  @Test
+  public void watermarkDoesNotChangeWhenToFewSampleRecords()
+      throws IOException, TransientKinesisException {
+    final long timestampMs = 1000L;
+
+    KinesisRecord firstRecord = prepareRecordMockWithArrivalTimestamp(timestampMs);
+    OngoingStubbing<CustomOptional<KinesisRecord>> firstIteratorStubbing =
+        when(firstIterator.next()).thenReturn(CustomOptional.of(firstRecord));
+    for (int i = 0; i < 5; i++) {
+      KinesisRecord record = prepareRecordMockWithArrivalTimestamp(timestampMs + i);
+      firstIteratorStubbing = firstIteratorStubbing.thenReturn(CustomOptional.of(record));
+    }
+    firstIteratorStubbing.thenReturn(CustomOptional.<KinesisRecord>absent());
+    when(secondIterator.next()).thenReturn(CustomOptional.<KinesisRecord>absent());
+
+    reader.start();
```

Comment:
Consider using the common pattern:

    for (boolean more = reader.start(); more; more = reader.advance())